tensorflow/python/ops/sparse_ops.py

```diff
@@ -2392,18 +2394,39 @@ def sparse_tensor_dense_matmul(sp_a,
       `return A*B`
   """
   # pylint: enable=line-too-long
-  sp_a = _convert_to_sparse_tensor(sp_a)
-  with ops.name_scope(name, "SparseTensorDenseMatMul",
-                      [sp_a.indices, sp_a.values, b]) as name:
-    b = ops.convert_to_tensor(b, name="b")
-    return gen_sparse_ops.sparse_tensor_dense_mat_mul(
-        a_indices=sp_a.indices,
-        a_values=sp_a.values,
-        a_shape=sp_a.dense_shape,
-        b=b,
-        adjoint_a=adjoint_a,
-        adjoint_b=adjoint_b)
 
+  if isinstance(b, sparse_tensor.SparseTensor) \
+          or isinstance(b, sparse_tensor.SparseTensorValue):
+
+    if adjoint_a == True and adjoint_b == False:
+      return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a,
+                                                            adjoint_a=True,
+                                                            adjoint_b=False))
+    elif adjoint_a == False and adjoint_b == True:
+      return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a,
+                                                            adjoint_a=False,
+                                                            adjoint_b=True))
+    elif adjoint_a == False and adjoint_b == False:
+      return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a,
+                                                            adjoint_a=True,
+                                                            adjoint_b=True))
+    elif adjoint_a == True and adjoint_b == True:
+      return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a,
+                                                            adjoint_a=False,
+                                                            adjoint_b=False))
```

Comment:
Actually, we need you to fix this. Is it intentional that the branches are changed here? (testing for both `andjoint_* == False` but setting them to `True` when calling the function)

If they have to be the same, why not just use

```
return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a, adjoin_a, adjoin_b)
```

Reply:
Don't forget about this change please. Is this intentional or a typo?

Reply:
Yes, we have to reverse the `adjoint_*` flags when you reverse the order of the multiply. 

To explain:

We have functionality for CD where C is sparse. To get the product AB where B is sparse, we calculate:

`AB = (B.T A.T).T `

This above format puts its back in the form that we have, namely, CD where C is sparse. Here, C now represents B.T, and D represents A.T.

So calculating AB where B is sparse requires the highest level call to have `adjoint_* = False`, while the sub-call, with the order swapped, requires them both to be true. That subcall also has to wrapped with the `transpose` op as you noted. 

All of this is to say that I think this is the most succinct way I can write the above logic, and it passes the tests to boot.

Does this make sense? Am I missing something obvious?

Reply:
Okay to resolve this segment?

Reply:
Oh, I just saw you're flipping the arguments. Now it makes more sense.

How about rewriting the entire code from line 2401 to 2417 like:

```python
    # We can do C * D where C is sparse but if we want to do A * B when
    # B is sparse we have to transpose. But AB = (B'A')' so we have to transpose.
    if adjoint_a != adjoint_b:
￼      return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a, adjoint_a, adjoint_b))
￼    else:
￼      return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a,
￼                                                            adjoint_a= not adjoin_a,
￼                                                            adjoint_b=not adjoin_b))
```

Comment probably needs rewrite, but code will be clearer, I think

Reply:
That looks great. Thanks for that!